### PR TITLE
Block Library: Add features to the Post Excerpt block.

### LINKED ISSUE #19698: New Block: Post Excerpt
As indicated in #15623, there will be a suite of Post blocks that make up the Post template.

This is the Post Excerpt block.

### Features

- Allow editing of the word count.
- Allow direct editing of the "Read more" link.

### Mockups

**Selected** (grays out the uneditable text)

Screen Shot 2020-01-15 at 8 20 33 PM

**Unselected** (text resumes normal color)

Screen Shot 2020-01-15 at 8 20 42 PM

### Questions

- Should we show a placeholder screen that allows the selection of a particular post, category, or tag?
- Should this block inherit a parent block's settings like category type, or tag?

### PR DESCRIPTION
**Pull request by epiqueras**
Closes #19698

## Description

This PR implements the design from #19698.

## How to test this?

Verify all the features from #19698 work as expected with a Post Excerpt block inserted in a regular post.

## Screenshots

gif

## Types of Changes

*New Feature*: The Post Excerpt block now has a more rich editing experience and allows you to customize the excerpt length and read more link directly.

## Checklist:

- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/post-excerpt/block.json [hunk 1]
```diff
@@ -1,4 +1,17 @@
 {
 	"name": "core/post-excerpt",
-	"category": "layout"
+	"category": "layout",
+	"attributes": {
+		"excerptLength": {
```
Comment (youknowriad) [inline packages/block-library/src/post-excerpt/block.json]: why not just "lenght"?

File: packages/block-library/src/post-excerpt/block.json [hunk 2]
```diff
@@ -1,4 +1,17 @@
...
+			"type": "number",
+			"default": 55
+		},
+		"moreText": {
+			"type": "string"
+		},
+		"showMoreOnNewLine": {
```
Comment (youknowriad) [inline packages/block-library/src/post-excerpt/block.json]: Should this be an attribute or just something a theme can tweak with CSS?

File: packages/block-library/src/post-excerpt/block.json (hunk 1 above)
Comment (epiqueras) [reply packages/block-library/src/post-excerpt/block.json]: It's clearer that it refers to the excerpt part. The block also has the read more text and could evolve to have other things in the future.

File: packages/block-library/src/post-excerpt/block.json (hunk 2 above)
Comment (epiqueras) [reply packages/block-library/src/post-excerpt/block.json]: Making it an attribute allows us to leverage different markup.

**Comment by paaljoachim** [comment]
For instance adding a post excerpt block on the post itself would mean that this specific block would not show up on the frontend when viewing the post. 
Will there be some kind of visual difference between the excerpt block and the rest of the post layout?

Meaning any block such as the post excerpt block that does not show up in the frontend post/page layout but has another specific feature could perhaps have a visual marker in the backend. To give it a clear difference between blocks used in the post/page layout and blocks that have a specific purpose not included it in the layout.

**Comment by epiqueras** [comment]
>For instance adding a post excerpt block on the post itself would mean that this specific block would not show up on the frontend when viewing the post.

It does render the excerpt for that post.

In any case, limiting things like this is what `settings.parent` is for.

File: packages/block-library/src/post-excerpt/block.json (hunk 1 above)
Comment (mtias) [reply packages/block-library/src/post-excerpt/block.json]: This should probably be `wordCount` or `wordLength`.

**Comment by karmatosed** [comment]
> Will there be some kind of visual difference between the excerpt block and the rest of the post layout?

I think this might be where CSS could be used. I could see exploring style variations but that feels like down to the theme almost.

File: packages/block-library/src/post-excerpt/edit.js [hunk 3]
```diff
@@ -2,16 +2,96 @@
...
+	const [ , , { raw: rawPostContent } ] = useEntityProp(
+		'postType',
+		'post',
+		'content'
+	);
+	return useMemo( () => {
+		if ( ! rawPostContent ) {
+			return '';
+		}
+		const excerptElement = document.createElement( 'div' );
```
Comment (MarcoZehe) [inline packages/block-library/src/post-excerpt/edit.js]: Is this the equivalent of the p element you use later in the PHP file? In other words, is this going to be used as a paragraph with inner text and a link, or is this a wrapper that will contain a p and something else? If the former, it should be a p element instead of a div, too.

Comment (epiqueras) [reply packages/block-library/src/post-excerpt/edit.js]: It's not rendered anywhere. It's just for parsing the HTML into a transformable text string.

**Review by draganescu — COMMENTED** [review]
Great work here :) One thing I think should be changed is the fact that `usePostContentExcerpt` has a custom logic that trims and cuts the post content in order to get the excerpt preview, while the server side `render_block_core_post_excerpt` uses the normal `get_the_excerpt`.

This will lead to things showing up in the editor which will not show up in the front.

As far as I can tell, I may be wrong, the REST API returns an element `['excerpt']['raw']` which is the result of some filters including `wp_trim_excerpt`. This filter/function takes care of generating an excerpt from the post data, so we shouldn't need to redo this logic. Also, the best thing it does, is to call `strip_shortcodes` and `excerpt_remove_blocks` which avoids bumping into issues like #19418

**Comment by epiqueras** [comment]
>This will lead to things showing up in the editor which will not show up in the front.

Do you have an example? They should be aligned.

>This filter/function takes care of generating an excerpt from the post data, so we shouldn't need to redo this logic.

Yes, we do because we are dynamically changing the excerpt length in the client and we wish to preview it while we do so.

>Also, the best thing it does, is to call strip_shortcodes and excerpt_remove_blocks which avoids bumping into issues like #19418

That's why I use rendered post content.

**Comment by draganescu** [comment]
> Do you have an example? They should be aligned.

Again, I am not 100% sure, but it appears that there are different sets of operations happening on each of the filters used for content or excerpt:

[default-filters.php#L172](https://github.com/WordPress/wordpress-develop/blob/c1abc8203a711e383f606301b3979c4a3f49718b/src/wp-includes/default-filters.php#L172)

```
add_filter( 'the_content', 'do_blocks', 9 );
add_filter( 'the_content', 'wptexturize' );
add_filter( 'the_content', 'convert_smilies', 20 );
add_filter( 'the_content', 'wpautop' );
add_filter( 'the_content', 'shortcode_unautop' );
add_filter( 'the_content', 'prepend_attachment' );
add_filter( 'the_content', 'wp_make_content_images_responsive' );

add_filter( 'the_excerpt', 'wptexturize' );
add_filter( 'the_excerpt', 'convert_smilies' );
add_filter( 'the_excerpt', 'convert_chars' );
add_filter( 'the_excerpt', 'wpautop' );
add_filter( 'the_excerpt', 'shortcode_unautop' );
add_filter( 'get_the_excerpt', 'wp_trim_excerpt', 10, 2 );
```

Most notably the_excerpt does not `do_blocks`, instead it does `excerpt_remove_blocks` in `wp_trim_excerpt`. On the other hand the REST controller for posts applies `the_content` filter on `$data['content']['rendered']`.

**Comment by epiqueras** [comment]
I meant in this PR. A scenario where it breaks.

I am taking the same approach the latest posts block takes and it seems to
work well.

wp_trim_excerpt also applies the_content.

excerpt_remove_blocks just removes blocks that won’t contribute to the
excerpt like images. This is achieved in the client by using innerHTML and
extracting text content only.

**Comment by draganescu** [comment]
So I have tried to reproduce the bug in #19418 and it worked:

Screenshot 2020-01-27 at 19 37 02

The point in the image above is that the excerpt now contains a link, and it should in fact be empty. To reproduce this:

1. add a post with an embed at the start
2. in that post add more content
3. use the excerpt block in that post or in single
4. you will see by default the link of the embed, which should be stripped

**Comment by epiqueras** [comment]
I think the correct fix here is not rendering that URL as visible text like that.

**Comment by epiqueras** [comment]
What's blocking this?

File: packages/block-library/src/post-excerpt/block.json (hunk 1 above)
Comment (youknowriad) [reply packages/block-library/src/post-excerpt/block.json]: wordCount sounds like the most accurate to me. Let's change it.

File: packages/block-library/src/post-excerpt/edit.js [hunk 4]
```diff
@@ -2,16 +2,96 @@
...
+}
+
+function PostExcerptEditor( {
+	attributes: { excerptLength, moreText, showMoreOnNewLine },
+	setAttributes,
+	isSelected,
+} ) {
 	const [ excerpt, setExcerpt ] = useEntityProp( 'postType', 'post', 'excerpt' );
-	return <PlainText value={ excerpt } onChange={ setExcerpt } />;
+	const postContentExcerpt = usePostContentExcerpt( excerptLength );
```
Comment (youknowriad) [inline packages/block-library/src/post-excerpt/edit.js]: Is this because the backend defaults to the post content if the excerpt is empty?

File: packages/block-library/src/post-excerpt/edit.js [hunk 5]
```diff
@@ -2,16 +2,96 @@
...
 import { useEntityProp, useEntityId } from '@wordpress/core-data';
-import { PlainText } from '@wordpress/block-editor';
+import { useMemo } from '@wordpress/element';
+import { InspectorControls, RichText } from '@wordpress/block-editor';
+import { PanelBody, RangeControl, ToggleControl } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
 
-function PostExcerptDisplay() {
+function usePostContentExcerpt( excerptLength ) {
+	const [ , , { raw: rawPostContent } ] = useEntityProp(
```
Comment (youknowriad) [inline packages/block-library/src/post-excerpt/edit.js]: Why do we have three arguments, is the first one normalized or something like that?

File: packages/block-library/src/post-excerpt/edit.js [hunk 6]
```diff
@@ -2,16 +2,96 @@
...
+				className={
+					! showMoreOnNewLine && 'wp-block-post-excerpt__excerpt is-inline'
+				}
+				placeholder={ postContentExcerpt }
+				value={ excerpt || ( isSelected ? '' : postContentExcerpt ) }
+				onChange={ setExcerpt }
+				keepPlaceholderOnFocus
+			/>
+			{ ! showMoreOnNewLine && ' ' }
+			<RichText
```
Comment (youknowriad) [inline packages/block-library/src/post-excerpt/edit.js]: Should this be inside a paragraph if "showMoreOnNewLine" is true (to match the backend)

File: packages/block-library/src/post-excerpt/block.json (hunk 1 above)
Comment (epiqueras) [reply packages/block-library/src/post-excerpt/block.json]: I disagree because of this being a "max length" and not a word count but I don't think it's a big deal so I've updated it.

File: packages/block-library/src/post-excerpt/edit.js (hunk 4 above)
Comment (epiqueras) [reply packages/block-library/src/post-excerpt/edit.js]: Yes, and you can't control the max length dynamically.

File: packages/block-library/src/post-excerpt/edit.js (hunk 5 above)
Comment (epiqueras) [reply packages/block-library/src/post-excerpt/edit.js]: The first one is the edited value, yes.

File: packages/block-library/src/post-excerpt/edit.js (hunk 6 above)
Comment (epiqueras) [reply packages/block-library/src/post-excerpt/edit.js]: Yes, updated.

File: lib/demo-block-templates/single.html [hunk 7]
```diff
@@ -0,0 +1,2 @@
+<!-- wp:post-title /-->
+<!-- wp:post-content /-->
```
Comment (youknowriad) [inline lib/demo-block-templates/single.html]: Seems like this is an unrelated change, do we want it or should we remove it from the PR?

**Review by youknowriad — APPROVED** [review]

File: lib/demo-block-templates/single.html (hunk 7 above)
Comment (epiqueras) [reply lib/demo-block-templates/single.html]: Oops, deleted.

File: packages/block-library/src/post-excerpt/edit.js [hunk 8]
```diff
@@ -2,20 +2,118 @@
...
 }
 
-export default function PostExcerptEdit() {
+export default function PostExcerptEdit( {
+	attributes,
+	setAttributes,
+	isSelected,
+} ) {
 	if ( ! useEntityId( 'postType', 'post' ) ) {
 		return 'Post Excerpt Placeholder';
```
Comment (markhowellsmead) [inline packages/block-library/src/post-excerpt/edit.js]: This text cannot be translated if it's defined statically.

File: packages/block-library/src/post-excerpt/index.php [hunk 9]
```diff
@@ -8,14 +8,39 @@
...
+
+	$filter_excerpt_length = function() use ( $attributes ) {
+		return isset( $attributes['wordCount'] ) ? $attributes['wordCount'] : 55;
+	};
+	add_filter(
+		'excerpt_length',
+		$filter_excerpt_length
+	);
+
+	$output = '<p>' . get_the_excerpt( $post );
```
Comment (markhowellsmead) [inline packages/block-library/src/post-excerpt/index.php]: This element should have a class name, so that it can be targeted by Theme and Plugin developers.
